### ISSUE TITLE
get_pair in the function add_liquidity_native may not return the correct value

### ISSUE THREAD

**Issue by fonstack** [hats-finance/AlephZeroAMM-0x0d88a9ece90994ecb3ba704730819d71c139f60f#15]
Github username: @JJtheAndroid
Twitter username: --
Severity: minor

**Description**\

get_pair in the function remove_liquidity_native may not return the correct value

**Attack Scenario**\

The function remove_liquidity_native calls the function get_pair to retrieve the address of a token pair

```
/// Returns address of a `Pair` contract instance (if exists) for
        /// `(token_0, token_1)` pair registered in `factory` Factory instance.
        #[inline]
        fn get_pair(
            &self,
            token_0: AccountId,
            token_1: AccountId,
        ) -> Result<AccountId, RouterError> {
            self.factory_ref()
                .get_pair(token_0, token_1)
                .ok_or(RouterError::PairNotFound)
        }
```

[https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/router/lib.rs#L42-L53](https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/router/lib.rs#L42-L53)

Pay attention to this particular line

```
 .get_pair(token_0, token_1)
```

Now let's take a look at the line for add_liquidity_native

```
 let pair_contract = self.get_pair(token, wnative)?;
```

[https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/router/lib.rs#L277](https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/router/lib.rs#L277)

add_liquidity_native assumes that a wnative pair will always be token 1 in a pair, but that may not necessarily be the case since token pair is defined as the following.

```
  #[ink(message)]
        fn create_pair(
            &mut self,
            token_0: AccountId,
            token_1: AccountId,
        ) -> Result<AccountId, FactoryError> {

```

[https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/factory/lib.rs#L110-L121](https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/factory/lib.rs#L110-L121)

There can be a case where the wnative is token0 instead of token1, which would not be detected by the get_pair function

**Attachments**

1. **Proof of Concept (PoC) File**

[https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/factory/lib.rs#L110-L121](https://github.com/Cardinal-Cryptography/common-amm/blob/bf4e48e3257894dcc8e6ab359321d1406533ad8b/amm/contracts/factory/lib.rs#L110-L121)

**Comment by deuszx** [comment]
When we [create_pair](https://github.com/hats-finance/AlephZeroAMM-0x0d88a9ece90994ecb3ba704730819d71c139f60f/blob/0a7264d707aea51b559a1bf94448681b59660f6a/amm/contracts/factory/lib.rs#L117-L121) we first sort the tokens. It is true that function expects the caller to sort the tokens as well. We could do it in the contract but we don't to save gas. There's a general rule of thumb that we try to use - if something doesn't need to be done in the contract it's not, to safe gas fees.

**Comment by JJtheAndroid** [comment]
> When we [create_pair](https://github.com/hats-finance/AlephZeroAMM-0x0d88a9ece90994ecb3ba704730819d71c139f60f/blob/0a7264d707aea51b559a1bf94448681b59660f6a/amm/contracts/factory/lib.rs#L117-L121) we first sort the tokens. It is true that function expects the caller to sort the tokens as well. We could do it in the contract but we don't to save gas. There's a general rule of thumb that we try to use - if something doesn't need to be done in the contract it's not, to safe gas fees.

That is part of the issue. The second part is that there is no assurance that wnative will always be token1 even with the sorting mechanism. For example, what is stopping something like this from happening? 

``` 
fn initialize_works() {
            let wnative = AccountId::from([0x03; 32]);
            let token_1 = AccountId::from([0x04; 32]);
            let pair = PairContract::new(wnative, token_1);
            assert_eq!(pair.get_token_0(), wnative);
            assert_eq!(pair.get_token_1(), token_1);
        }
```
Here the sorting mechanism works as intended, and wnative is still not token1

**Comment by coreggon11** [comment]
@JJtheAndroid the Factory only works with tokens created by the Factory. [Here](https://github.com/hats-finance/AlephZeroAMM-0x0d88a9ece90994ecb3ba704730819d71c139f60f/blob/0a7264d707aea51b559a1bf94448681b59660f6a/amm/contracts/factory/lib.rs#L131-L134) the `token0` and `token1` will be mapped to the pair contract created by the factory. So the order in which you are retrieving the pair contract does not matter (since both `(token0,token1)` and `(token1,token0)` point to the same `pair`).  Creating a pair outside of the contract (what you have described) is not a part of the AMM, since it will not be registered in the Factory.

Correct me if I'm wrong:)

**Comment by JJtheAndroid** [comment]
> @JJtheAndroid the Factory only works with tokens created by the Factory. [Here](https://github.com/hats-finance/AlephZeroAMM-0x0d88a9ece90994ecb3ba704730819d71c139f60f/blob/0a7264d707aea51b559a1bf94448681b59660f6a/amm/contracts/factory/lib.rs#L131-L134) the `token0` and `token1` will be mapped to the pair contract created by the factory. So the order in which you are retrieving the pair contract does not matter (since both `(token0,token1)` and `(token1,token0)` point to the same `pair`). Creating a pair outside of the contract (what you have described) is not a part of the AMM, since it will not be registered in the Factory.
> 
> Correct me if I'm wrong:)

That is not correct. In AMMs the factory creates the pair contract, it does not create tokens. The tokens that the pair contract represents (token0, token1) can be any token. I only created that quick test in the pair contract file because it was easier. I could have done it in the factory contract file and the result would still be the same. 

You are right that both token pair point to the same contract. However this is not handled properly as shown here and the original report.

https://github.com/hats-finance/AlephZeroAMM-0x0d88a9ece90994ecb3ba704730819d71c139f60f/blob/b097173adc9966bcbed72c6a4f1b50fcc52fe0ef/amm/contracts/factory/lib.rs#L172-L176

The function still expects a certain order of the token, so the mapping doesn't matter


If they checked both sides like this (pseudo code), then I think you would be correct

```
        fn get_pair(&self, token_0: AccountId, token_1: AccountId) -> Option<AccountId> {
            self.get_pair.get((token_0, token_1))
        if  self.get_pair.get((token_0, token_1)) returns an error {
           self.get_pair.get((token_1, token_0))
           }
```

**Comment by coreggon11** [comment]
I meant to write Factory works with **pair** tokens created by the Factory.

Still, if I create a pair of 0x03 and 0x04, get_pair(0x03, 0x04) would return the same pair as get_pair(0x04, 0x03), as factory maps both these tuples to the same pair (as shown in the code snippet provided earlier).

**Comment by deuszx** [comment]
@coreggon11 is correct - `Pair` instances live "in the void" - i.e. you can create multiple `Pair` instances for the same token and they are independent of each other. They need to be "registered" in a `Factory` instance (which is done implicitly when one adds liquidity to a `Pair` through the corresponding `Router`) for them to be "visible" to the `Router`'s users. As @coreggon11 pointed out, upon creation of a `Pair` it's registered under both addresses - `(a,b)` and `(b,a)` so the `get_pair` will always return.

**Comment by deuszx** [comment]
Thank you for participation. After carefully reviewing the submission we've concluded that this issue is INVALID.

We hope you participate in the future audits of ink!.
